### PR TITLE
Add permissions for stable API

### PR DESCRIPTION
**Pull request by jhtimmins**
Related: https://github.com/apache/airflow/issues/8112

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: airflow/api_connexion/security.py
Comment (houqp) [inline airflow/api_connexion/security.py]: should we combine @security.requires_access and @security.requires_authentication into one decorator instead? will there be a case where one would only use require_access without requires_authentication? this doesn't make sense semantically right?

File: airflow/www/security.py
Comment (houqp) [inline airflow/www/security.py]: some of the resources here like ImportError doesn't have edit/delete/create endpoints right?

File: airflow/www/security.py
Comment (houqp) [inline airflow/www/security.py]: is it expected to keep this comment here?

File: tests/test_utils/api_connexion_utils.py
Comment (mik-laj) [inline tests/test_utils/api_connexion_utils.py:38]: What do you think about extending the create_user method to accept a list of permissions?
```python
def create_user(app, username, role=None, permissions=None):
    appbuilder = app.appbuilder
    current_role = None
    if role and permission:
    	raise IllegalArgument("role and permission are muttually exclussive")
    elif role:
	    current_role = appbuilder.sm.find_role(role)
	elif permission:
		ranadom_suffix = get_random_string()
	    current_role = create_role(app, get_random_string("test-role-{ranadom_suffix}"))

    tester = appbuilder.sm.find_user(username=username)
    if not tester:
        appbuilder.sm.add_user(
            username=username,
            first_name=username,
            last_name=username,
            email=f"{username}@fab.org",
            role=current_role,
            password=username,
        )
```
This will simplify testing as we won't have to worry so much about side effects as each role will only be used once. This will also explicitly associate the user with a specific permission set.

**Comment by mik-laj** [comment]
Can you add some docs? We have docs for access control, but that doesn't describe the permission in detail.
https://airflow.readthedocs.io/en/latest/security/access-control.html 

Here is similar documentation on access control on other services.
https://cloud.google.com/composer/docs/how-to/access-control
https://cloud.google.com/workflows/docs/access-control

File: airflow/www/security.py
Comment (jhtimmins) [reply airflow/www/security.py]: Good point @houqp. I've updated it so that the resource is created dynamically when the decorator is called on the function.

File: airflow/www/security.py
Comment (jhtimmins) [reply airflow/www/security.py]: Removed

File: airflow/api_connexion/security.py
Comment (jhtimmins) [reply airflow/api_connexion/security.py]: @houqp we could do that, but there are times when authentication is needed, but we don't tie it to a specific permission. Such as the Health and Configs endpoints. We could add a semi-arbitrary permission for those cases (`('Health', 'can_read')`), or similar with reading configs. Or we could allow `requires_access()` to work without any permissions, in which case it only requires authentication.

What do you think? @mik-laj curious about your thoughts as well.

Comment (houqp) [reply airflow/api_connexion/security.py]: For endpoints that need to be guarded by authentication but not authorization, one solution we can adopt is to pass in an empty permissions list, or `None`, then in the decorator skip the role check entirely if that's the case.

Comment (jhtimmins) [reply airflow/api_connexion/security.py]: yeah that isnt particularly difficult, but I think it complicates the interface somewhat. I think I'll just create resources that aren't based on a specific model. So 'Health' or 'Configuration', just to keep the format consistent

File: airflow/api_connexion/endpoints/health_endpoint.py
Comment (mik-laj) [inline airflow/api_connexion/endpoints/health_endpoint.py]: This endpoint does not need authorization as it will be used to verify that the instance is working properly even if no account has been created yet.


The Kubernetes API also don't have authorization for health endpoint.
```
$ ENDPOINT=$(gcloud container clusters describe airflow-cluster --zone us-central1 --format='value(endpoint)')
$ curl --insecure https://$ENDPOINT/readyz?verbose
[+]ping ok
[+]log ok
[+]etcd ok
[+]poststarthook/generic-apiserver-start-informers ok
[+]poststarthook/start-apiextensions-informers ok
[+]poststarthook/start-apiextensions-controllers ok
[+]poststarthook/crd-informer-synced ok
[+]poststarthook/bootstrap-controller ok
[+]poststarthook/rbac/bootstrap-roles ok
[+]poststarthook/scheduling/bootstrap-system-priority-classes ok
[+]SSH Tunnel Check ok
[+]poststarthook/ca-registration ok
[+]poststarthook/start-kube-apiserver-admission-initializer ok
[+]poststarthook/start-kube-aggregator-informers ok
[+]poststarthook/apiservice-registration-controller ok
[+]poststarthook/apiservice-status-available-controller ok
[+]poststarthook/kube-apiserver-autoregistration ok
[+]autoregister-completion ok
[+]poststarthook/apiservice-openapi-controller ok
[+]shutdown ok
healthz check passed
```

File: airflow/api_connexion/endpoints/version_endpoint.py
Comment (mik-laj) [inline airflow/api_connexion/endpoints/version_endpoint.py]: This will allow the client's behavior to be changed before authorization is performed.  I am able to imagine that other authorization methods may be required depending on the version.

Kubernetes API don't have aaccesss control on /version endpoint
```
$ ENDPOINT=$(gcloud container clusters describe airflow-cluster --zone us-central1 --format='value(endpoint)')
$ curl --insecure https://$ENDPOINT/version
{
  "major": "1",
  "minor": "16+",
  "gitVersion": "v1.16.13-gke.1",
  "gitCommit": "688c6543aa4b285355723f100302d80431e411cc",
  "gitTreeState": "clean",
  "buildDate": "2020-07-21T02:37:26Z",
  "goVersion": "go1.13.9b4",
  "compiler": "gc",
  "platform": "linux/amd64"
}
```

File: airflow/api_connexion/security.py
Comment (mik-laj) [reply airflow/api_connexion/security.py]: Config endpoint should have authentication and authorization.
Health and version endpoint don't need any protection. There are no sensitive data there and for operational reasons it is better not to add them. This will greatly facilitate monitoring, service autodiscovery and API client development.

Comment (jhtimmins) [reply airflow/api_connexion/security.py]: Thanks. Good to know. I've removed the Health and Version authorization.

File: airflow/www/app.py
Comment (ashb) [inline airflow/www/app.py]: This won't do anything if we dont' pass any permissions in.

File: airflow/api_connexion/security.py
Comment (houqp) [inline airflow/api_connexion/security.py]: nitpick, could you add type annotations to this function?

File: airflow/api_connexion/security.py
Comment (houqp) [inline airflow/api_connexion/security.py]: shouldn't we doing an early return here instead of creating an empty permissions list?

File: airflow/www/security.py
Comment (houqp) [inline airflow/www/security.py]: same as my previous comment, seems like we can just do early return here.

**Review by houqp — COMMENTED** [review]
left 3 minor comments, the rest looks good to me 👍

File: airflow/www/security.py
Comment (ashb) [inline airflow/www/security.py:166]: Where is `can_dag_read` still used?

File: airflow/api_connexion/security.py
Comment (houqp) [inline airflow/api_connexion/security.py]: could you annotate the return type as well?

```suggestion
) -> None:
```

**Review by houqp — COMMENTED** [review]
also need to handle merge conflicts with master :)

File: airflow/api_connexion/security.py
Comment (ashb) [inline airflow/api_connexion/security.py:58]: Should we check dag level permissions for DagRun/TaskInstance too?

Perhaps anywhere we have a `dag_id` named parameter?

Comment (ashb) [reply airflow/api_connexion/security.py:58]: Ah, no, this is fine, we just need to add `('can_read', 'Dag')` to the listed permissions in dagrun end points.

File: airflow/www/security.py
Comment (jhtimmins) [reply airflow/www/security.py:166]: Currently, the UI is still using `can_dag_read`. When I backport the new permissions I can remove this.

**Comment by jhtimmins** [comment]
@ashb @mik-laj @houqp Ok I think I've addressed or included all of your edits.

**Comment by mik-laj** [comment]
I think it is worth adding a description of the required permissions for each endpoint to the OpenAPI specification. What do you think about it? Currently, this information is only available in code.

**Comment by mik-laj** [comment]
Alternatively, you can add the tables to the documentation and that will be enough for me too.
Screenshot 2020-09-05 at 03 07 44

**Review by houqp — APPROVED** [review]
LGTM, please fix CI steps.

**Comment by jhtimmins** [comment]
@mik-laj @ashb Ok I've added docs that show the permissions for each endpoint. https://github.com/apache/airflow/pull/10594/files#diff-1da5e874639ba7d6631e109fd81e9dadR130

This is also a good place to review the permissions I've added to the endpoints, in case y'all disagree with my choices. Thanks!

File: docs/security/access-control.rst
Comment (ashb) [inline docs/security/access-control.rst:155]: I don't think we need Task as a separate permission entity -- it's not separately useful enough I don't think

Comment (ashb) [reply docs/security/access-control.rst:155]:
```suggestion
/dags/{dag_id}/tasks                                                               GET    Dag.can_read
/dags/{dag_id}/tasks/{task_id}                                                     GET    Dag.can_read
```

File: docs/security/access-control.rst
Comment (ashb) [inline docs/security/access-control.rst]:
```suggestion
/dags/{dag_id}/dagRuns/{dag_run_id}/taskInstances/{task_id}/links                  GET    Dag.can_read, DagRun.can_read, Task.can_read, TaskInstance.can_read
```

File: docs/security/access-control.rst
Comment (ashb) [inline docs/security/access-control.rst]:
```suggestion
/dags/{dag_id}/dagRuns/{dag_run_id}/taskInstances/list                             POST   Dag.can_read, DagRun.can_read, Task.can_read, TaskInstance.can_read
```

File: docs/security/access-control.rst
Comment (ashb) [inline docs/security/access-control.rst]:
```suggestion
```

File: docs/security/access-control.rst
Comment (kaxil) [inline docs/security/access-control.rst]: This might be a rebase deletion! Or was this intentional?

File: docs/security/access-control.rst
Comment (houqp) [reply docs/security/access-control.rst:155]: I agree, it's better to start without it and add it later if it turned out to be something that's actually needed.

**Comment by jhtimmins** [comment]
@mik-laj Ok this is ready to merge, if you wouldn't mind approving/merging

**Comment by mik-laj** [comment]
@jhtimmins I am starting to look at it and there seems to be one problem so please hold off the merge.

**Comment by mik-laj** [comment]
If there are 3 DAGs in the database: ``DAG_A``, ``DAG_B``, and you have permissions - ``[can_read, DAG_A]``, then you will send a request - `GET /api/v1/dags`, what DAGs will you get? I guess you will either get information on all DAGs, or you will not get information on any.

Your database query should be different depending on the accessible DAGs for the user.
https://github.com/apache/airflow/blob/eaa49b2257913c34b15408a14e445f6106e691ee/airflow/www/views.py#L287
https://github.com/apache/airflow/blob/eaa49b2257913c34b15408a14e445f6106e691ee/airflow/www/views.py#L305-L306

**Comment by jhtimmins** [comment]
Good catch @mik-laj. I've updated the authorization code to properly handle cases where a user has access to some dags but not all. The endpoint is also updated to only return the proper results. https://github.com/apache/airflow/pull/10594/files#diff-9982141783adf996be12cffcf84d6fe5R59

**Comment by mik-laj** [comment]
Have you looked at this issue in other views as well?  For example: Endpoint - "List XCom entries" allows specifying `~` as the dag_id, dag_run_id, task_id to retrieve XCOM entries for all DAGs, DAG runs and task instances.

This address URL returns all entries for all DAG in the database.
https://localhostot/api/v1/dags/~/dagRuns/~/taskInstances/~/xcomEntries

File: airflow/models/dag.py
Comment (ashb) [inline airflow/models/dag.py]: This feels out of place here in the DAG class -- it leaks FAB into Airflow core, which so far has almost entirely managed to mot be tied together.

I think this would be better if we kept that separation, so is it possible to keep this in the SecurityManager class?

File: airflow/models/dag.py
Comment (ashb) [inline airflow/models/dag.py]: Won't this also include things like "can_index on Airflow", resulting in `WHERE dag_id IN (..., "Airflow", ...)` etc?

File: airflow/models/dag.py
Comment (ashb) [inline airflow/models/dag.py]: This is the new "all_dags" permission right?

Comment (jhtimmins) [reply airflow/models/dag.py]: Affirmative

File: airflow/models/dag.py
Comment (jhtimmins) [reply airflow/models/dag.py]: @ashb It's certainly possible. I was deciding between letting flask code leak into Airflow core and putting core code into the security manager. I chose the former, but I can switch back to the latter.

File: airflow/models/dag.py
Comment (jhtimmins) [reply airflow/models/dag.py]: This will look at that permission, but the check on 1692 will only filter on permissions with an action (`can_read`), matching the specified action from the function args.

However, it will grab all resources with matching actions. So it would grab `Airflow.can_read`, if `action == 'can_read'`. I figure that was fine though, since it isn't an issue unless users are naming dags `Airflow`, `TaskInstance`, etc.

**Comment by jhtimmins** [comment]
@mik-laj @ashb Ok I've incorporated both of your requested changes. Mind letting me know if anything is missing? If it looks good, feel free to merge.

File: airflow/api_connexion/endpoints/dag_run_endpoint.py
Comment (ashb) [inline airflow/api_connexion/endpoints/dag_run_endpoint.py:125]: This would filter before any date limits are applied, and thus would get an incorrect count.

File: airflow/api_connexion/endpoints/dag_run_endpoint.py
Comment (ashb) [inline airflow/api_connexion/endpoints/dag_run_endpoint.py:180]: What was the reason for removing the session here?

File: airflow/api_connexion/endpoints/xcom_endpoint.py
Comment (ashb) [inline airflow/api_connexion/endpoints/xcom_endpoint.py]:
```suggestion
        query.join(DR, and_(XCom.dag_id == DR.dag_id, XCom.execution_date == DR.execution_date))
```

This is not your PR causing this -- but I also think that this is missing a `query = ` at the start -- I don't _think_ `.join()` mutates the query, but returns a new one. I could be wrong on this. (Separate PR to fix it if this _doesn't_ work as is)

File: airflow/api_connexion/endpoints/xcom_endpoint.py
Comment (ashb) [inline airflow/api_connexion/endpoints/xcom_endpoint.py:61]: When there's a single dag_id the permissions decorator will validate access to that dag, correct?

**Comment by ashb** [comment]
@mik-laj Any final comments? James has addressed the per-dag access controls now.

There's one or two small things here to fix, but if you have no further comments I'll merge this tomorrow lunchtime.

File: airflow/api_connexion/endpoints/dag_run_endpoint.py
Comment (jhtimmins) [reply airflow/api_connexion/endpoints/dag_run_endpoint.py:125]: Good catch. I think the original behavior included a bug. @mik-laj can you confirm that the count should be taken after filtering by date?

Comment (mik-laj) [reply airflow/api_connexion/endpoints/dag_run_endpoint.py:125]: `total_entries` should return the number of elements that are available on all the page of the current request.

File: airflow/api_connexion/endpoints/dag_run_endpoint.py
Comment (jhtimmins) [reply airflow/api_connexion/endpoints/dag_run_endpoint.py:180]: It has to do with how the count is getting made, as seen in your previous comment. It used to use `session`, but it doesn't need to now.

File: docs/security/access-control.rst
Comment (mik-laj) [inline docs/security/access-control.rst]: The beginning of this paragraph is repeated.

File: airflow/api_connexion/endpoints/dag_run_endpoint.py
Comment (jhtimmins) [reply airflow/api_connexion/endpoints/dag_run_endpoint.py:125]: Actually, I think we should maintain the existing behavior here and do the count prior to the date filtering. Just to keep scope limited. If we decide that the count should take place after the date filtering, let's make that behavioral change in a separate PR, just to reduce scope for this one somewhat.

File: airflow/api_connexion/endpoints/xcom_endpoint.py
Comment (jhtimmins) [reply airflow/api_connexion/endpoints/xcom_endpoint.py:61]: Correct

File: airflow/api_connexion/endpoints/dag_run_endpoint.py
Comment (mik-laj) [reply airflow/api_connexion/endpoints/dag_run_endpoint.py:125]: Can you create a issues about it?

**Comment by mik-laj** [comment]
@ashb  I don't have any major comments. It looks good now.

File: airflow/api_connexion/endpoints/xcom_endpoint.py
Comment (jhtimmins) [reply airflow/api_connexion/endpoints/xcom_endpoint.py]: I've made the update, and created an issue to add a test for this

https://github.com/apache/airflow/issues/11073

File: airflow/api_connexion/endpoints/dag_run_endpoint.py
Comment (jhtimmins) [reply airflow/api_connexion/endpoints/dag_run_endpoint.py:125]: I've created an issue to change how the total_entries are counted. #11074

**Pull request merged**
